docs/spec/spec.md

```diff
@@ -0,0 +1,416 @@
+| byte                                  | byte           | INT1                  |
+| byte[]                                | byte[]         | RAW, LRAW, BYTE       |
+| decimal                               | BigDecimal     | DEC, QUAN, CURR       |
+| int                                   | BigInteger     | NUM                   |
+| record (with/without rest fields)     | JCoStructure   | STRUCTURE             |
+| record[] (with/without rest fields)   | JCoTable       | TABLE                 |
+
+### 2.2 Listener
+
+#### 2.2.1 Configurations
+
+The Listener supports two types of configurations, `ServerConfig` and `AdvancedConfig`, to facilitate communication with the SAP system by receiving IDocs and handling events.
+
+##### 2.2.1.1 ServerConfig
+
+The `ServerConfig` type represents the configuration details needed to create an IDoc connection.
+
+```ballerina
+public type ServerConfig record {
+    # The gateway host (jco.server.gwhost)
+    string gwhost;
+    # The gateway service (jco.server.gwserv)
+    string gwserv;
+    # The program ID (jco.server.progid)
+    string progid;
+};
+```
+
+**Example Configuration:**
+
+```ballerina
+configurable jco:ServerConfig configs = ?;
+listener jco:Listener idocListener = new (configs);
+```
+
+Config.toml file:
+
+```toml
+[configs]
+gwhost = "gwhost"
+gwserv = "sapgw00"
+progid = "progid"
+```
+
+##### 2.2.1.2 AdvancedConfig
+
+If the user needs to specify more detailed configurations beyond `ServerConfig`—for example, to register a destination with the listener—they can use `AdvancedConfig`. The `AdvancedConfig` type is a map that holds any SAP JCo configurations accepted by the SAP server provider.
+
+```ballerina
+public type AdvancedConfig map<string>;
+```
+
+**Example Configuration:**
+
+```ballerina
+configurable jco:AdvancedConfig configs = ?;
+listener jco:Listener idocListener = new (configs);
+```
+
+Config.toml file:
+
+```toml
+[configs]
+"jco.server.gwhost" = "gwhost"
+"jco.server.gwserv" = "sapgw00"
+"jco.server.repository_destination" = "SAMPLE_DESTINATION"
+"jco.client.ashost" = "host"
+"jco.client.sysnr" = "00"
+"jco.client.client" = "000"
+"jco.client.user" = "user"
+"jco.client.passwd" = "password"
+```
+Here's how the updated section could look, incorporating the provided code:
+
+#### 2.2.2 Functions
+
+This section describes how to attach, detach, start, and stop the listener, along with initializing it for listening to iDoc messages.
+
+- To attach a service to the iDoc listener, the `attach` function can be used.
+
+```ballerina
+# Attaches a `Service` to the iDoc listener.
+# ```
+# check idocListener.attach(service);
+# ```
+#
+# + `s` - The service instance to be attached.
+# + `name` - The name of the service (optional).
+# + `return` - An error if the attachment fails or else `()`.
+public isolated function attach(Service s, string[]|string? name = ()) returns error? = @java:Method {
```

Comment:
This question is not directly regarding the spec. The usual practice is to define a module level error type and use it in anywhere there can be an error return. Any reason to deviate from that here ?

Reply:
Yes, fixed with https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/57.